libraries/developer_account.rb

```diff
@@ -5,32 +5,35 @@ module MacOS
   class DeveloperAccount
     attr_reader :credentials
 
-    def initialize(data_bag_retrieval, node_credential_attributes, download_url)
+    def initialize(apple_id_hash, download_url)
       if download_url.empty?
-        developer_id = find_apple_id(data_bag_retrieval, node_credential_attributes)
+        developer_id = validate_apple_id(apple_id_hash)
         @credentials = {
-          XCODE_INSTALL_USER: developer_id['apple_id'],
-          XCODE_INSTALL_PASSWORD: developer_id['password'],
+          XCODE_INSTALL_USER: developer_id[:username],
+          XCODE_INSTALL_PASSWORD: developer_id[:password],
         }
         authenticate_with_apple(@credentials)
       end
     end
 
-    def authenticate_with_apple(credentials)
-      shell_out!(XCVersion.update, env: credentials)
-    end
-
-    def find_apple_id(data_bag_retrieval, node_credential_attributes)
-      if node_credential_attributes
+    def validate_apple_id(apple_id_hash)
```

Comment:
Instead of the duplication here, maybe try to catch a `KeyError` using `Hash#fetch`:

```ruby
if apple_id_hash.values.any?
    {
      username: apple_id_hash[:username],
      password: apple_id_hash[:password],
     }
else
    begin
        {
          username: ENV.fetch('XCODE_INSTALL_USER'),
          password: ENV.fetch('XCODE_INSTALL_PASSWORD'),
        }
    rescue KeyError
        raise('Developer credentials not supplied, and a URL was not provided for Xcode!')
    end
end
```

Also, not sure why the error message says a URL was not provided for Xcode? This looks like it's only related to user credentials.

Reply:
Not sure which duplication this resolves, you have two sets of potential assignment to the `username` and `password` keys. I actually think the `if`/`elsif`/`else` is more straightforward than nesting `begin`/`rescue` under `else`.

You don't have to supply developer credentials if there is a URL for an Xcode xip provided. This raises if neither was provided.